### PR TITLE
Add API to update pr headBranch api

### PR DESCRIPTION
**Pull request by a1012112796**
As title.

### CODE REVIEW

File: routers/api/v1/api.go [hunk 1]
```diff
@@ -800,6 +800,7 @@ func RegisterRoutes(m *macaron.Macaron) {
 							Patch(reqToken(), reqRepoWriter(models.UnitTypePullRequests), bind(api.EditPullRequestOption{}), repo.EditPullRequest)
 						m.Get(".diff", repo.DownloadPullDiff)
 						m.Get(".patch", repo.DownloadPullPatch)
+						m.Post("/update", repo.UpdatePullRequest)
```
Comment (lunny) [inline routers/api/v1/api.go]: reqToken()

**Comment by lunny** [comment]
Otherwise LGTM

File: routers/api/v1/api.go (hunk 1 above)
Comment (zeripath) [reply routers/api/v1/api.go]:
```suggestion
						m.Post("/update", reqToken(), repo.UpdatePullRequest)
```

**Review by zeripath — APPROVED** [review]

File: routers/api/v1/repo/pull.go [hunk 2]
```diff
@@ -968,3 +968,97 @@ func parseCompareInfo(ctx *context.APIContext, form api.CreatePullRequestOption)
...
+		if models.IsErrPullRequestNotExist(err) {
+			ctx.NotFound()
+		} else {
+			ctx.Error(http.StatusInternalServerError, "GetPullRequestByIndex", err)
+		}
+		return
+	}
+
+	if pr.HasMerged {
+		ctx.Error(http.StatusNotFound, "UpdatePullRequest", err)
```
Comment (6543) [inline routers/api/v1/repo/pull.go]: I would return 422 - since PR exist

File: routers/api/v1/repo/pull.go [hunk 3]
```diff
@@ -968,3 +968,97 @@ func parseCompareInfo(ctx *context.APIContext, form api.CreatePullRequestOption)
...
+		return
+	}
+
+	if err = pr.LoadIssue(); err != nil {
+		ctx.Error(http.StatusInternalServerError, "LoadIssue", err)
+		return
+	}
+
+	if pr.Issue.IsClosed {
+		ctx.Error(http.StatusNotFound, "UpdatePullRequest", err)
```
Comment (6543) [inline routers/api/v1/repo/pull.go]: If PR is not merged but closed it still exist too :)

File: routers/api/v1/repo/pull.go [hunk 4]
```diff
@@ -968,3 +968,97 @@ func parseCompareInfo(ctx *context.APIContext, form api.CreatePullRequestOption)
...
+		return
+	}
+
+	if !allowedUpdate {
+		ctx.Status(http.StatusMethodNotAllowed)
+		return
+	}
+
+	// default merge commit message
+	message := fmt.Sprintf("Merge branch '%s' into %s", pr.BaseBranch, pr.HeadBranch)
```
Comment (6543) [inline routers/api/v1/repo/pull.go:1054]: [OPTIONAL] you could add an optional option in body for the message

**Comment by 6543** [comment]
not that bad :) - thanks @a1012112796

**Review by 6543 — CHANGES_REQUESTED** [review]
block

File: routers/api/v1/repo/pull.go (hunk 4 above)
Comment (a1012112796) [reply routers/api/v1/repo/pull.go:1054]: I'd like it to be consistent with UI, so I won't change it for now. Thanks.

**Comment by zeripath** [comment]
@6543 are you happy now?

**Review by 6543 — APPROVED** [review]
I am :)

**Comment by zeripath** [comment]
make lg-tm work

**Pull request merged**